In [1]:
import librosa

In [2]:
import librosa.display

In [3]:
# Importing the Keras libraries and packages
from keras.models import Sequential
from keras.layers import Convolution2D
from keras.layers import MaxPooling2D
from keras.layers import Flatten
from keras.layers import Dense


Using TensorFlow backend.


### Here is a helper function for converting .wav sound files into melspectrogram
I have used **melspectrogram** here beacuse I felt it was more visually intuitive than others like mfcc although i would encourage you to try out others!!
But in order for this to work we have to make some folders to pass in the appropriate path:
 sound data<br>
-------------            train<br>
------------------                  dogs<br>
------------------                 cats<br>
-------------            test<br>
------------------                dogs<br>
------------------                 cats<br>
                 

In [ ]:
def save_fig_dog(loc,train_or_test,cat_or_dog,start,stop):
    if cat_or_dog=="dog":
        for i in range(start,stop):
            a=loc+"dog_barking_{}.wav".format(i)
            y,sr=librosa.load(a)
            M=librosa.feature.melspectrogram(y=y)
            plt.plot(M)
            plt.savefig('sound data/{}/dogs/dog{}'.format(train_or_test,i),dpi=400, bbox_inches='tight',pad_inches=0)
            plt.close()
        print("done")
    elif cat_or_dog=="cat":
        for i in range(start,stop):
            a=loc+"cat_{}.wav".format(i)
            y,sr=librosa.load(a)
            M=librosa.feature.melspectrogram(y=y)
            plt.plot(M)
            M=librosa.feature.melspectrogram(y=y)
            plt.savefig('sound data/{}/cats/cat{}'.format(train_or_test,i),dpi=400, bbox_inches='tight',pad_inches=0)
            plt.close()
        print("done")
    else:
        print("error")   

In [4]:
#Initialising the CNN
classifier = Sequential()

# Step 1 - Convolution
classifier.add(Convolution2D(32, 3, 3, input_shape = (64,64, 3), activation = 'relu'))

# Step 2 - Pooling
classifier.add(MaxPooling2D(pool_size = (2, 2)))

# Adding a second convolutional layer
classifier.add(Convolution2D(32, 3, 3, activation = 'relu'))
classifier.add(MaxPooling2D(pool_size = (2, 2)))

# Step 3 - Flattening
classifier.add(Flatten())

# Step 4 - Full connection
classifier.add(Dense(output_dim = 128, activation = 'relu'))
classifier.add(Dense(output_dim = 1, activation = 'sigmoid'))

# Compiling the CNN
classifier.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])

W0205 01:14:01.517208 13136 module_wrapper.py:139] From D:\Anaconda\lib\site-packages\keras\backend\tensorflow_backend.py:74: The name tf.get_default_graph is deprecated. Please use tf.compat.v1.get_default_graph instead.

D:\Anaconda\lib\site-packages\ipykernel_launcher.py:5: UserWarning: Update your `Conv2D` call to the Keras 2 API: `Conv2D(32, (3, 3), input_shape=(64, 64, 3..., activation="relu")`
  """
W0205 01:14:01.520201 13136 module_wrapper.py:139] From D:\Anaconda\lib\site-packages\keras\backend\tensorflow_backend.py:517: The name tf.placeholder is deprecated. Please use tf.compat.v1.placeholder instead.

W0205 01:14:01.524221 13136 module_wrapper.py:139] From D:\Anaconda\lib\site-packages\keras\backend\tensorflow_backend.py:4138: The name tf.random_uniform is deprecated. Please use tf.random.uniform instead.

W0205 01:14:01.538184 13136 module_wrapper.py:139] From D:\Anaconda\lib\site-packages\keras\backend\tensorflow_backend.py:3976: The name tf.nn.max_pool is deprecated. Pl

In [5]:
from keras.preprocessing.image import ImageDataGenerator

In [6]:
from keras.preprocessing.image import ImageDataGenerator

# Data Augmentation

train_datagen = ImageDataGenerator(rescale = 1./255)

test_datagen = ImageDataGenerator(rescale = 1./255)

training_set = train_datagen.flow_from_directory('sound data/train',
                                                 target_size = (64,64),
                                                 batch_size = 16,
                                                 class_mode = 'binary')

test_set = test_datagen.flow_from_directory('sound data/test',
                                            target_size = (64,64),
                                            batch_size = 16,
                                            class_mode = 'binary')



Found 207 images belonging to 2 classes.
Found 69 images belonging to 2 classes.


In [7]:
classifier.fit_generator(training_set,# the training set
                         samples_per_epoch = 207,
                         nb_epoch = 9,# number of epochs 
                         validation_data = test_set,# the test set
                         nb_val_samples = 69)

D:\Anaconda\lib\site-packages\ipykernel_launcher.py:5: UserWarning: The semantics of the Keras 2 argument `steps_per_epoch` is not the same as the Keras 1 argument `samples_per_epoch`. `steps_per_epoch` is the number of batches to draw from the generator at each epoch. Basically steps_per_epoch = samples_per_epoch/batch_size. Similarly `nb_val_samples`->`validation_steps` and `val_samples`->`steps` arguments have changed. Update your method calls accordingly.
  """
D:\Anaconda\lib\site-packages\ipykernel_launcher.py:5: UserWarning: Update your `fit_generator` call to the Keras 2 API: `fit_generator(<keras_pre..., validation_data=<keras_pre..., steps_per_epoch=12, epochs=9, validation_steps=69)`
  """
W0205 01:14:01.994136 13136 module_wrapper.py:139] From D:\Anaconda\lib\site-packages\keras\backend\tensorflow_backend.py:986: The name tf.assign_add is deprecated. Please use tf.compat.v1.assign_add instead.

W0205 01:14:02.085653 13136 module_wrapper.py:139] From D:\Anaconda\lib\site-pac

Epoch 1/9


W0205 01:14:02.718940 13136 module_wrapper.py:139] From D:\Anaconda\lib\site-packages\keras\backend\tensorflow_backend.py:174: The name tf.get_default_session is deprecated. Please use tf.compat.v1.get_default_session instead.

W0205 01:14:02.718940 13136 module_wrapper.py:139] From D:\Anaconda\lib\site-packages\keras\backend\tensorflow_backend.py:181: The name tf.ConfigProto is deprecated. Please use tf.compat.v1.ConfigProto instead.

W0205 01:14:02.926058 13136 module_wrapper.py:139] From D:\Anaconda\lib\site-packages\keras\backend\tensorflow_backend.py:190: The name tf.global_variables is deprecated. Please use tf.compat.v1.global_variables instead.

W0205 01:14:02.926058 13136 module_wrapper.py:139] From D:\Anaconda\lib\site-packages\keras\backend\tensorflow_backend.py:199: The name tf.is_variable_initialized is deprecated. Please use tf.compat.v1.is_variable_initialized instead.

W0205 01:14:02.991540 13136 module_wrapper.py:139] From D:\Anaconda\lib\site-packages\keras\backend\te

12/12 [==============================] - 32s 3s/step - loss: 0.8547 - acc: 0.4531 - val_loss: 0.6775 - val_acc: 0.5931
Epoch 2/9
12/12 [==============================] - 30s 3s/step - loss: 0.6702 - acc: 0.6128 - val_loss: 0.6542 - val_acc: 0.5937
Epoch 3/9
12/12 [==============================] - 30s 3s/step - loss: 0.6407 - acc: 0.6646 - val_loss: 0.5845 - val_acc: 0.8537
Epoch 4/9
12/12 [==============================] - 30s 3s/step - loss: 0.4869 - acc: 0.7857 - val_loss: 0.4130 - val_acc: 0.9000
Epoch 5/9
12/12 [==============================] - 30s 3s/step - loss: 0.3316 - acc: 0.8792 - val_loss: 0.3415 - val_acc: 0.9137
Epoch 6/9
12/12 [==============================] - 31s 3s/step - loss: 0.2934 - acc: 0.8952 - val_loss: 0.3198 - val_acc: 0.8845
Epoch 7/9
12/12 [==============================] - 30s 3s/step - loss: 0.2773 - acc: 0.9000 - val_loss: 0.3158 - val_acc: 0.8695
Epoch 8/9
12/12 [==============================] - 30s 3s/step - loss: 0.2093 - acc: 0.9261 - val_loss: 0.4

## The training Accuracy is 93.7% and validation accuracy is 85% which is again pretty good considering the simplicity of the Neural Network Architecture used here